falcon/cluster/cluster.py

```diff
@@ -547,10 +701,392 @@ def _get_cluster_medoids(
+) -> bool:
+    """
+    Check if a value is an outlier.
+
+    Parameters
+    ----------
+    value : float
+        The value to check.
+    median : float
+        The median of the values.
+    std : float
+        The standard deviation of the values.
+    n_min : float
+        The number of standard deviations below the median.
+    n_max : float
+        The number of standard deviations above the median.
+
+    Returns
+    -------
+    bool
+        True if the value is an outlier, False otherwise
+    """
+    return (value < median - n_min * std) or (value > median + n_max * std)
+
+
+@nb.njit(cache=True)
+def _average_peaks(
+    bins_indices: List[int], bins_peaks: nb.typed.List
+) -> nb.typed.List:
+    """
+    Average the peaks in each bin.
+
+    Parameters
+    ----------
+    bins_indices : List[int]
+        The indices of the non-empty bins.
+    bins_peaks : nb.typed.List
+        The intensities for each bin.
+
+    Returns
+    -------
+    nb.typed.List
+        The average intensities for each non-empty bin.
+    """
+    avged_bins_peaks = np.empty(len(bins_indices), np.float32)
+
+    for i in range(len(bins_indices)):
+        intensities = bins_peaks[i]
+        if len(intensities) < 1:
+            continue
+        elif len(intensities) == 1:
+            avged_bins_peaks[i] = intensities[0]
         else:
-            medoids.append(idx_interval[start_i])
-    return np.asarray(medoids, dtype=np.int32)
+            avged_bins_peaks[i] = np.mean(intensities)
+
+    return avged_bins_peaks
+
+
+@nb.njit(cache=True)
+def _construct_average_spectrum(
+    bins_indices: List[int],
+    bins_peaks: nb.typed.List,
+    avg_precursor_mz: float,
+    charge: int,
+    min_mz: float,
+    bin_size: float,
+    avg_rt: float,
+    cluster: int,
+    cluster_size: int,
+) -> similarity.SpectrumTuple:
+    """
+    Construct the average spectrum from the binned spectra.
+
+    Parameters
+    ----------
+    bins_indices : List[int]
+        The indices of the non-empty bins.
+    bins_peaks : nb.typed.List
+        The intensities for each bin.
+    avg_precursor_mz : float
+        The average precursor m/z.
+    charge : int
+        The precursor charge.
+    min_mz : float
+        The minimum m/z to include in the vectors.
+    bin_size : float
+        The bin size in m/z used to divide the m/z range.
+    avg_rt : float
+        The average retention time.
+    cluster : int
+        The cluster label.
+    cluster_size : int
+        The number of spectra in the cluster.
+
+    Returns
+    -------
+    similarity.SpectrumTuple
+        The average spectrum.
+    """
+    mz = np.empty(len(bins_indices), np.float32)
+    intensity = np.empty(len(bins_indices), np.float32)
+
+    idx = 0
+    for bin_idx, avg_intensity in zip(bins_indices, bins_peaks):
+        # use the middle of the bin as the m/z value
+        mz[idx] = min_mz + (bin_idx * bin_size) + (bin_size / 2)
```

Comment:
**thought:** Is this what the algorithm from the paper does? I'd think that taking the average of the m/z values might be better. That would better capture if the m/z's are not evenly distributed over the bin.